### PR TITLE
Fix and refactor banner fixing

### PR DESCRIPTION
**Pull request by ZehMatt**
I slightly refactored how this works, it should be also easier to understand now. I also re-ordered to how things get fixed.

The first fix that is applied is iterating all BannerElement tiles looking for elements that share the same banner id, it will create for each duplicate element a new banner and copies the data from the shared banner.

The second fix that is applied is the banner position, the previous duplicate fixup keeps the current position, this step will iterate BannerElement elements and sets the correct position on each banner entry. The old fix only fixed those of duplicates, this now ensures that each banner has the correct position.

The last fix that is now applied is deleting banners that have no banner element, this previously happened before the duplicate fixup but this step did not consider that the entry might exist but has the position wrong, the previous step now ensures that this won't delete banners that aren't supposed to be deleted due the bad position.

This solves https://github.com/OpenRCT2/OpenRCT2/issues?q=is%3Aissue+is%3Aopen+BannerFixDuplicates

### CODE REVIEW

File: src/openrct2/world/Banner.cpp [hunk 1]
```diff
@@ -255,60 +285,83 @@ void BannerResetBrokenIndex()
...
-                        auto newBannerId = newBanner->id;
-
-                        *newBanner = *oldBanner;
-                        newBanner->id = newBannerId;
-                        newBanner->position = { x, y };
-                    }
-
-                    bannerElement->SetIndex(newBanner->id);
-                }
+        auto newBanner = CreateBanner();
```
Comment (frutiemax) [inline src/openrct2/world/Banner.cpp:319]: Put const auto*

File: src/openrct2/world/Banner.cpp [hunk 2]
```diff
@@ -255,60 +285,83 @@ void BannerResetBrokenIndex()
...
+}
+
+// Ensures that all banner entries have the correct position based on the element
+// that references the banner entry.
+static void BannerFixPositions(std::vector<BannerElementWithPos>& bannerElements)
+{
+    for (const auto& entry : bannerElements)
+    {
+        const auto index = entry.Element->GetIndex();
+        auto* banner = GetBanner(index);
```
Comment (frutiemax) [inline src/openrct2/world/Banner.cpp:344]: const auto*

**Review by frutiemax — COMMENTED** [review]
Minor things

File: src/openrct2/world/Banner.cpp (hunk 1 above)
Comment (ZehMatt) [reply src/openrct2/world/Banner.cpp:319]: I need to write into it, so it can't be const.

File: src/openrct2/world/Banner.cpp (hunk 2 above)
Comment (ZehMatt) [reply src/openrct2/world/Banner.cpp:344]: It will be modified, can't be const.

**Review by duncanspumpkin — APPROVED** [review]
Sounds sensible, not tested

**Review by Broxzier — APPROVED** [review]

**Comment by Broxzier** [comment]
Tested and didn't find any issues. Code seems fine. Right after accepting this PR. I realized this could do with a changelog entry.

**Review by Broxzier — CHANGES_REQUESTED** [review]
Changelog

**Review by Broxzier — APPROVED** [review]

**Pull request merged**